### PR TITLE
feat: calculate-hashes command to detect contract changes

### PR DESCRIPTION
**Pull request by benceharomi**
# What ❔
Created a `calculate-hashes` script that saves the hash of the preprocessed source code and bytecode of the system contracts.

## Why ❔
We should clearly see when bootloader (and account abstraction) hashes have changed.

We should have a file in the repo, that contains the current hashes (of both compiled and source codes - as for bootloader and AA - we actually do some preprocessing of the source code).

we could have a small tool in era-system-contracts, that computes the hash of the compiled file (using zksync_utils::bytecode::hash_bytecode) and also a hash of the source code - and compares it with the hashes that are in the file.

This way, whenever we do the change, the PR would also contain the changes to that file.

## Checklist
- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `yarn prettier:write` and `yarn lint:fix`.

### CODE REVIEW

File: SystemContractsHashes.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (StanislavBreadless) [inline SystemContractsHashes.json:1]: Maybe expand this PR to also cover hashes of all system contracts?

File: scripts/calculate-hashes.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      "Failed to calculate Bootloader hashes."
+    ),
+    defaultAA: getHashes(
+      defaultAASourceCode,
+      defaultAABytecode,
+      "Failed to calculate DefaultAccount hashes."
+    ),
+  };
+
+  const newSystemContractsHashes = JSON.stringify(
```
Comment (StanislavBreadless) [inline scripts/calculate-hashes.ts:88]: are you sure that the order is preserved? i.e. you are stringifying an object, I am not sure that JS guarantees fixed ordering for objects

File: scripts/calculate-hashes.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  const oldSystemContractsHashes = fs.readFileSync(OUTPUT_FILE_PATH, "utf8");
+
+  if (oldSystemContractsHashes === newSystemContractsHashes) {
+    console.log(
+      "Calculated hashes match the hashes in the SystemContractsHashes.json file."
+    );
+    return;
+  } else if (checkOnly) {
+    console.error(
+      "Calculated hashes differ from the hashes in the SystemContractsHashes.json file. Exiting..."
```
Comment (StanislavBreadless) [inline scripts/calculate-hashes.ts:103]: better to also output the differing hashes.

File: scripts/calculate-hashes.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+};
+
+const getHashes = (
+  sourceCode: string,
+  bytecode: string,
+  errorMessage: string
+): Hashes => {
+  try {
+    return {
+      sourceCodeHash: ethers.utils.sha256(sourceCode),
```
Comment (mm-zk) [inline scripts/calculate-hashes.ts:47]: please comment why we use different hash method for code and for bytecode

File: scripts/calculate-hashes.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    throw new Error(errorMessage);
+  }
+};
+
+const main = async () => {
+  const checkOnly = process.argv.includes("--check-only");
+
+  const bootloaderSourceCode = readFileAsHexString(
+    BOOTLOADER_SOURCE_CODE_PATH,
+    "Failed to read Bootloader source code. Make sure to run `yarn build-yul` before you run this script!"
```
Comment (mm-zk) [inline scripts/calculate-hashes.ts:60]: please also print  the CODE_PATH value in the error message (this usually makes it a lot easier to debug).

here and in other lines below

File: scripts/calculate-hashes.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  if (oldSystemContractsHashes === newSystemContractsHashes) {
+    console.log(
+      "Calculated hashes match the hashes in the SystemContractsHashes.json file."
+    );
+    return;
+  } else if (checkOnly) {
+    console.error(
+      "Calculated hashes differ from the hashes in the SystemContractsHashes.json file. Exiting..."
+    );
```
Comment (mm-zk) [inline scripts/calculate-hashes.ts:104]: and print the message - that you should run the 'yarn calculate-hashes' to update the file.

**Review by mm-zk — APPROVED** [review]

File: SystemContractsHashes.json (hunk 1 above)
Comment (benceharomi) [reply SystemContractsHashes.json:1]: other contracts added

File: scripts/calculate-hashes.ts (hunk 2 above)
Comment (benceharomi) [reply scripts/calculate-hashes.ts]: changed to lodash's `isEqual`

File: scripts/calculate-hashes.ts (hunk 3 above)
Comment (benceharomi) [reply scripts/calculate-hashes.ts]: Differnce is printed

File: scripts/calculate-hashes.ts (hunk 4 above)
Comment (benceharomi) [reply scripts/calculate-hashes.ts]: Comment added

File: scripts/calculate-hashes.ts (hunk 6 above)
Comment (benceharomi) [reply scripts/calculate-hashes.ts:335]: Message added

File: scripts/calculate-hashes.ts (hunk 5 above)
Comment (benceharomi) [reply scripts/calculate-hashes.ts]: path is printed

**Review by nateawelch — APPROVED** [review]

File: scripts/calculate-hashes.ts [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+type ContractDetails = {
+  contractName: string;
+  sourceCodePath: string;
+  bytecodePath: string;
+};
+
+const SYSTEM_CONTRACTS_DETAILS: ContractDetails[] = [
+  // contracts dir sol
+  {
+    contractName: "AccountCodeStorage",
```
[the remark this answers is not in the document]
Comment (nateawelch) [reply scripts/calculate-hashes.ts]: Agreed, I've spent a few hours before debugging similar issues where I just assumed contracts would be picked up automatically

[the remark this answers is not in the document]
Comment (benceharomi) [reply scripts/calculate-hashes.ts]: I had some concerns about doing this automatically, but I can figure out a way that works.

File: scripts/calculate-hashes.ts [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+    sourceCodePath: "contracts/AccountCodeStorage.sol",
+    bytecodePath:
+      "artifacts-zk/cache-zk/solpp-generated-contracts/AccountCodeStorage.sol/AccountCodeStorage.json",
```
[the remark this answers is not in the document]
Comment (benceharomi) [reply scripts/calculate-hashes.ts]: changed

**Pull request merged**